control-plane/agents/src/bin/core/volume/specs.rs

```diff
@@ -1115,13 +1141,24 @@ impl ResourceSpecsLocked {
             .await
     }
 
-    /// Remove volume by its `id`
+    /// Remove volume by its `id`.
     pub(super) fn remove_volume(&self, id: &VolumeId) {
         let mut specs = self.write();
         specs.volumes.remove(id);
     }
 
-    /// Get or Create the resourced VolumeSpec for the given request
+    /// Remove volume group by its `id` only if the volume list becomes empty.
+    pub(super) fn remove_volume_group(&self, id: &VolumeId, vg_id: &String) {
+        let mut specs = self.write();
+        if let Some(vg_spec) = specs.volume_groups.get(vg_id) {
+            vg_spec.lock().remove(id);
+            if vg_spec.lock().is_empty() {
+                specs.volume_groups.remove(vg_id);
+            }
```

Comment:
is this racy? After checking `is_empty` another one might be added? Maybe keep vg_spec locked while we remove it?

Reply:
> We're not using the vg in the other scenarios where we add replicas, will that be on another PR?

Yes, I will add the other operations, in the upcoming PRs. Till this point was mainly the structural part without any impact.